### PR TITLE
Make Kedro-Viz shareable via a hosted URL

### LINKED ISSUE #1116: Shareable URLs for Kedro-Viz
## Description 

This idea focuses on creating a way for users to share versions of their current **Kedro-Viz pipeline visualisation** for others to explore. We've often seen this request for [interactive sharing of a pipeline visualisation arise from technical "translators" (people who are technical enough to run `kedro-viz`](https://confluence.quantumblack.com/pages/viewpage.action?pageId=303337178). This group wanted ways to share their pipeline visualisation with people who could not make a copy of the Kedro project, install Kedro-Viz and run Kedro-Viz to see the pipeline (normally other senior stakeholders). 

This item is a high-priority issue because there is no other way to open up Kedro-Viz to non-technical users unless we make it possible for them to access Kedro-Viz without a technical or "translator" user. 

Users currently solve the problem of sharing a version of their current Kedro-Viz pipeline by: 

 - Download a PNG version of their Kedro-Viz pipeline
 - Screenshotting a version of their Kedro-Viz pipeline
 - Creating a hosted Kedro-Viz (one user did this through a Dash app)
 - And in the past, they used a link to Studio, an internal product, to see an interactive version of their plots

I've focused this issue on pipeline visualisation sharing but, we will also need to think about what this means for Collaborative Experiment Tracking users. 

### Based on exporting something from `kedro-viz`
A suggestion for a possible user experience based on the CLI could be:

1. User sets up the pipeline in a Kedro project
2. User installs Kedro-Viz with `pip install kedro-viz` from the CLI
3. User runs `kedro viz export` 
4. User receives a URL which can be shared with other team members

A suggestion for a possible user experience on the UI could be: 

1. User sets up the pipeline in a Kedro project
2. User installs Kedro-Viz with `pip install kedro-viz` from the CLI
3. User runs `kedro viz` and opens up the UI
4. User clicks on an icon to generate the URL
5. A pop-up appears with the generated link 
6. User receives a URL which can be shared with other team members

### Based on interacting with the project GitHub repo
Another user experience could be that a user: 
 - Heads to a website
 - Sees a form which asks for a link to a GitHub repo and what branch needs to be visualised
 - User inserts link to GitHub repo and which branch 
 - User gets a generated link

> I was mucking around on Powerpoint and made this, it's similar to the way that [Streamlit](https://streamlit.io/cloud) works. Please do not take this as a design but let it inspire you. This design also has security features because it's behind a login.

Kapture 2023-02-23 at 09 48 12

## Other user considerations

Additional research on this feature has shown additional requirements that users will have: 
 - The security of shared links will be an essential requirement; i.e. we'll need to think of a workflow where access to the visualisation is restricted to certain people
 - We'll need to figure out how to host the URLs too; our users have struggled to do this on their own 
 - One user flagged a need to ship a certain branch in their git-workflow i.e. the team might have broken the pipeline in `develop` but the pipeline visualisation still works on `main`
 - One user wanted an ability to see a changelog (or see what's changed between the pipeline) on Studio people used to be able to see timestamped versions of Kedro-Viz when a new version was submitted but not a delta between pipelines
 
And additionally, there are outstanding questions like: 
 - What happens if a user updates the pipeline? Does this mean they need to send a new link or will their current one update?
 - What happens if Collaborative Experiment Tracking is enabled?
 - Users have also spoken about being able to edit parameters and re-do the run. What would happen if we enabled #1279?

### PR DESCRIPTION
**Pull request by tynandebold**
## Description

Closes #1116 

## Development notes

This PR enables Viz to be published and hosted on AWS S3.

For the frontend, most changes exist in `shareable-url-modal.js`. On the deployed version, we show a timestamp for context, which comes from the `shareable-url-metadata.js` file. 

A new feature hint for the Sharable URLs button is added in `feature-hints-content.js`. I've also updated `feature-hints.js` to remove that feature hint from the deployed version, since it doesn't make sense for the user to see it there.

## QA notes

> [!IMPORTANT]  
> You don't need to review any Python files in this PR (nothing inside `/package`)!!! That review will be done in #1498.

To get setup to test this locally:

- Install `fsspec` version `2023.9.0` or greater.
- Ensure you're running `kedro` version `18.2.0` or greater.
- You'll need access to an AWS S3 bucket and know the following information:
    - Access key
    - Secret access key
    - Bucket region
    - Bucket name
- Enable static website hosting on your S3 bucket and then make the objects public so you'll be able to view Viz when it's deployed 
- Set your access keys in your terminal, and then run `make run`.
- Then run the frontend with `npm start`.
- The deploy button is in the lower-left, between the theme toggle and settings icon.

## Checklist

- [x] Read the [contributing](/CONTRIBUTING.md) guidelines
- [x] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [x] Updated the documentation to reflect the code changes
- [x] Added new entries to the `RELEASE.md` file
- [x] Added tests to cover my changes

<a href="https://gitpod.io/#https://github.com/kedro-org/kedro-viz/pull/1487"></a>

### CODE REVIEW

File: src/components/shareable-url-modal/shareable-url-modal.js [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+    setDeploymentState('loading');
+    setIsLoading(true);
+
+    try {
+      const request = await fetch('/api/deploy', {
+        headers: {
+          'Content-Type': 'application/json',
+        },
+        method: 'POST',
+        body: JSON.stringify(inputValues),
```
Comment (rashidakanchwala) [inline src/components/shareable-url-modal/shareable-url-modal.js:47]: @tynandebold  -- you will have to do the conversion from camel case to snake case and send it as input_values to the backend. 

I looked up online, and that's usually the approach bcoz FE has linting issues is we don't use camelCase and BE has same if we don't use snake_case

File: src/components/flowchart-wrapper/flowchart-wrapper.scss [hunk 2]
```diff
@@ -78,3 +78,11 @@ $sidebar-toolbar-width-open: variables.$sidebar-width-open +
     transform: translate(-50px, -140%);
   }
 }
+
+.shareable-url-button .button__btn {
+  // background: variables.$white-0;
+  // color: variables.$black-900;
```
Comment (jitu5) [inline src/components/flowchart-wrapper/flowchart-wrapper.scss:84]: Nit: We can remove unwanted commented code.

File: src/components/shareable-url-modal/shareable-url-modal.js [hunk 3]
```diff
@@ -0,0 +1,309 @@
...
+  'eu-south-1',
+  'eu-west-3',
+  'eu-north-1',
+  'eu-south-2',
+  'eu-central-2',
+  'sa-east-1',
+  'me-south-1',
+  'me-central-1',
+  'il-central-1',
+];
```
Comment (jitu5) [inline src/components/shareable-url-modal/shareable-url-modal.js:50]: Nit: We can put this is in some common location, may be in utils?

**Comment by jitu5** [comment]
LGTM!! 👍 One or two optional nit picking comments.

**Review by jitu5 — APPROVED** [review]

**Comment by jitu5** [comment]
Few observation about shareable URL modal:
1. When the region dropdown is not selected, the bucket name field is empty and you click on the deploy button, input fields validation is missing. Also The current error message ("Something went wrong. Please try again later.") may need improvement, such as adjusting text placement or colour.
2. Can not close the region dropdown when clicking outside.
3. Modal is not closing no esc key press.(By @tynandebold).
4. As you can see bucket name input field seems like pre populated text and not as editable input field.
Screenshot 2023-09-21 at 5 40 33 p m
 
 @tynandebold @stephkaiser

File: src/components/shareable-url-modal/shareable-url-metadata.js [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+            Accept: 'application/json',
+          },
+        });
+        const response = await request.json();
+
+        if (request.ok) {
+          setMetadata(response);
+        }
+      } catch (error) {
+        console.log('deploy-viz-metadata fetch error: ', error);
```
Comment (vladimir-mck) [inline src/components/shareable-url-modal/shareable-url-metadata.js:26]: We should think of a better way to display an error or at least replace the `log` with `warn` or `error`.

**Review by vladimir-mck — APPROVED** [review]
A nit comment about error handling but otherwise LGTM

File: RELEASE.md [hunk 5]
```diff
@@ -8,10 +8,14 @@ Please follow the established format:
...
+- Make Kedro-Viz shareable via a hosted URL. (#1487)
+
 ## Bug fixes and other changes
 
-- Fix for Kedro Viz Connection Error (#1507)
-- Fix display of modular pipeline nodes that are associated with tags (#1542)
+- Fix for Kedro-Viz Connection Error. (#1507)
+- Fix display of modular pipeline nodes that are associated with tags. (#1542)
 
 # Release 6.5.1
```
Comment (ravi-kumar-pilla) [inline RELEASE.md:20]: Not sure if we want to remove this (6.5.1) release note here but commenting here before we merge it to main

File: package/kedro_viz/api/rest/router.py [hunk 6]
```diff
@@ -63,9 +63,7 @@ async def deploy_kedro_viz(input_values: S3DeployerConfiguration):
         )
     except Exception as exc:  # pragma: no cover
         logger.exception("Deploying Kedro Viz failed: %s ", exc)
-        return JSONResponse(
-            status_code=500, content={"message": "Failed to deploy Kedro Viz"}
-        )
+        return JSONResponse(status_code=500, content={"message": f"{exc}"})
```
Comment (ravi-kumar-pilla) [inline package/kedro_viz/api/rest/router.py:66]: Not sure if we want to return the exception to the user (will appear in the developer tools even if handled on FE).

File: src/components/shareable-url-modal/shareable-url-metadata.js [hunk 7]
```diff
@@ -0,0 +1,46 @@
+import { useEffect, useState } from 'react';
+import { isRunningLocally } from '../../utils';
+
+const ShareableUrlMetadata = () => {
+  const [metadata, setMetadata] = useState(null);
+
+  useEffect(() => {
+    if (isRunningLocally()) {
```
Comment (ravi-kumar-pilla) [inline src/components/shareable-url-modal/shareable-url-metadata.js:8]: can we have this check when the component is mounted like when it is used in the page instead of checking it here ?

File: src/config.js [hunk 8]
```diff
@@ -145,3 +145,36 @@ export const errorMessages = {
...
+  'eu-south-1',
+  'eu-west-3',
+  'eu-north-1',
+  'eu-south-2',
+  'eu-central-2',
+  'sa-east-1',
+  'me-south-1',
+  'me-central-1',
+  'il-central-1',
+];
```
Comment (ravi-kumar-pilla) [inline src/config.js:180]: I think this is fine for now. But, it would be nice if we can use some sort of library like boto3 in the backend and return the list of available regions instead of hardcoding in the FE. Since AWS regions might be changed over time.

**Review by ravi-kumar-pilla — COMMENTED** [review]
I have made few suggestions, none of them is a blocking change. Looks great !

File: package/kedro_viz/api/rest/router.py (hunk 6 above)
Comment (rashidakanchwala) [reply package/kedro_viz/api/rest/router.py:66]: I was thinking more on this. isn't it better to return something on the front-end on why the deploy failed for instance : S3 Credentials incorrect or S3FS not installed etc so the user knows how to fix things. 
presently, we don't do this on Kedro-viz, the user is expected to see the error on the terminal. So i am not sure how we want to tackle it here.

File: src/apollo/subscriptions.js [hunk 9]
```diff
@@ -1,13 +0,0 @@
-import gql from 'graphql-tag';
-
-/** subscribe to receive new runs */
-export const NEW_RUN_SUBSCRIPTION = gql`
```
Comment (rashidakanchwala) [inline src/apollo/subscriptions.js:4]: @tynandebold - Since we are removing this subscription feature from Experiment Tracking I think it should have it's own PR. So in the future, incase we want to add it back or we want to explore this again, there's history. Especially because based on this subscription feature we were planning other features like auto-reload on Collab Experiment Tracking and so on. 

Right now because it's part of Shareable-viz -- it is not obvious.

File: src/components/feature-hints/feature-hint-dot.js [hunk 10]
```diff
@@ -1,12 +1,11 @@
 import React, { useEffect, useState } from 'react';
 import { connect } from 'react-redux';
 
-import { featureHintsContent } from './feature-hints-content';
-
 import './feature-hints.scss';
 
 const FeatureHintDot = ({
   appState,
+  featureHintsContent,
```
Comment (rashidakanchwala) [inline src/components/feature-hints/feature-hint-dot.js:8]: I am confused again because there are a lot of featureHints changes in this PR which seem out of scope and there's no information in the description so I don't know have much context of why they are in the PR.

Comment (ravi-kumar-pilla) [reply src/components/feature-hints/feature-hint-dot.js:8]: +1

File: package/kedro_viz/api/rest/router.py (hunk 6 above)
Comment (ravi-kumar-pilla) [reply package/kedro_viz/api/rest/router.py:66]: it makes sense to return the exception to FE if we want FE to show the users exactly what is going wrong. So, I feel a better approach would be to show the users an error box on viz something like Deploy failed and an expandable div showing the exception details. I am not sure how a non-developer using shareable viz can sort this out (either in the UI or going through the dev tools).

File: src/components/shareable-url-modal/shareable-url-modal.js [hunk 11]
```diff
@@ -0,0 +1,336 @@
...
+import Input from '../ui/input';
+import LoadingIcon from '../icons/loading';
+import Modal from '../ui/modal';
+import MenuOption from '../ui/menu-option';
+
+import './shareable-url-modal.scss';
+
+const modalMessages = (status, info = '') => {
+  const messages = {
+    default:
```
Comment (rashidakanchwala) [inline src/components/shareable-url-modal/shareable-url-modal.js:22]: can the error messages be in the config.js?

File: src/components/flowchart-wrapper/flowchart-wrapper.js [hunk 12]
```diff
@@ -288,9 +296,11 @@ export const FlowChartWrapper = ({
           >
             <LoadingIcon visible={loading} />
           </div>
+          <ShareableUrlMetadata />
         </div>
         <ExportModal />
         <MetadataModal />
+        <ShareableUrlModal />
```
Comment (rashidakanchwala) [inline src/components/flowchart-wrapper/flowchart-wrapper.js:303]: could we do  
{ isRunningLocal() && <ShareableUrlModal /> here then we don't need to in the component?

File: src/components/shareable-url-modal/shareable-url-modal.js [hunk 13]
```diff
@@ -0,0 +1,336 @@
...
+    </Modal>
+  );
+};
+
+export const mapStateToProps = (state) => ({
+  visible: state.visible,
+});
+
+export const mapDispatchToProps = (dispatch) => ({
+  onToggle: (value) => {
```
Comment (rashidakanchwala) [inline src/components/shareable-url-modal/shareable-url-modal.js:331]: nit

```suggestion
  onToggleModal: (value) => {
```

File: src/components/shareable-url-modal/shareable-url-metadata.js (hunk 7 above)
Comment (tynandebold) [reply src/components/shareable-url-modal/shareable-url-metadata.js:8]: Updated.

File: src/components/feature-hints/feature-hint-dot.js (hunk 10 above)
Comment (tynandebold) [reply src/components/feature-hints/feature-hint-dot.js:8]: We want to highlight this new feature to the users, don't we? 

image

Did it not make sense when you checked out the branch and ran it locally?

File: src/components/shareable-url-modal/shareable-url-modal.js (hunk 13 above)
Comment (tynandebold) [reply src/components/shareable-url-modal/shareable-url-modal.js:331]: Updated.

File: src/components/flowchart-wrapper/flowchart-wrapper.js (hunk 12 above)
Comment (tynandebold) [reply src/components/flowchart-wrapper/flowchart-wrapper.js:303]: Updated.

File: src/components/shareable-url-modal/shareable-url-modal.js (hunk 11 above)
Comment (tynandebold) [reply src/components/shareable-url-modal/shareable-url-modal.js:22]: It doesn't feel like they belong there since they're so specific to this component.

File: src/apollo/subscriptions.js (hunk 9 above)
Comment (tynandebold) [reply src/apollo/subscriptions.js:4]: Fair enough.

Please review this today, if you can: https://github.com/kedro-org/kedro-viz/pull/1554

**Review by stephkaiser — APPROVED** [review]

File: src/components/feature-hints/feature-hints.js [hunk 14]
```diff
@@ -15,7 +16,10 @@ import './feature-hints.scss';
 
 const localStorageKeyShowHints = 'showFeatureHints';
 export const localStorageKeyFeatureHintsStep = 'featureHintStep';
-const numFeatureHints = featureHintsContent.length;
+const updatedFeatureHintsContent = isRunningLocally()
```
Comment (rashidakanchwala) [inline src/components/feature-hints/feature-hints.js:19]: This is not a blocker but just for clarity--  from what I understand 'Publish and share Kedro-viz' feature hint shouldn't show in the deployed version and hence we have to write this code. 

But this feature hint only gets shown when the user hovers over the 'deploy button' and there won't be any deploy button, so why do we still need to have this check here?

Comment (tynandebold) [reply src/components/feature-hints/feature-hints.js:19]: Because even if the indicator dot isn't there the content in the card would still exist.

File: src/components/shareable-url-modal/shareable-url-modal.js (hunk 11 above)
Comment (rashidakanchwala) [reply src/components/shareable-url-modal/shareable-url-modal.js:21]: Sure, ok with that. It's just easier to manage when all the hardcoded content is in one place.

**Review by rashidakanchwala — APPROVED** [review]
Thanks for this!

**Review by ravi-kumar-pilla — APPROVED** [review]
